plugin/trino-hive/src/test/java/io/trino/plugin/hive/optimizer/TestConnectorPushdownRulesWithHive.java

```diff
@@ -284,6 +286,87 @@ public void testColumnPruningProjectionPushdown()
         metastore.dropTable(new HiveIdentity(SESSION), SCHEMA_NAME, tableName, true);
     }
 
+    @Test
+    public void testPushdownWithDuplicateExpressions()
+    {
+        String tableName = "duplicate_expressions";
+        tester().getQueryRunner().execute(format(
+                "CREATE TABLE  %s (struct_of_bigint, just_bigint) AS SELECT cast(row(5, 6) AS row(a bigint, b bigint)) AS struct_of_int, 5 AS just_bigint WHERE false",
+                tableName));
+
+        PushProjectionIntoTableScan pushProjectionIntoTableScan = new PushProjectionIntoTableScan(
+                tester().getMetadata(),
+                tester().getTypeAnalyzer(),
+                new ScalarStatsCalculator(tester().getMetadata(), tester().getTypeAnalyzer()));
+
+        HiveTableHandle hiveTable = new HiveTableHandle(SCHEMA_NAME, tableName, ImmutableMap.of(), ImmutableList.of(), ImmutableList.of(), Optional.empty());
+        TableHandle table = new TableHandle(new CatalogName(HIVE_CATALOG_NAME), hiveTable, new HiveTransactionHandle(), Optional.empty());
+
+        HiveColumnHandle bigintColumn = createBaseColumn("just_bigint", 1, toHiveType(BIGINT), BIGINT, REGULAR, Optional.empty());
+        HiveColumnHandle partialColumn = new HiveColumnHandle(
+                "struct_of_bigint",
+                0,
+                toHiveType(ROW_TYPE),
+                ROW_TYPE,
+                Optional.of(new HiveColumnProjectionInfo(
+                        ImmutableList.of(0),
+                        ImmutableList.of("a"),
+                        toHiveType(BIGINT),
+                        BIGINT)),
+                REGULAR,
+                Optional.empty());
+
+        // Test projection pushdown with duplicate constants
```

Comment:
How about duplicating the column reference instead of the constant? Then in the result of the transformation you can see that the column reference was the expression which actually interacted with the connector. And that in turn proves that it had to go (both occurrences) through the NodeRef map which you fixed.

```
        // Test projection pushdown with duplicate column references
        tester().assertThat(pushProjectionIntoTableScan)
                .on(p -> {
                    SymbolReference column = p.symbol("just_bigint", BIGINT).toSymbolReference();
                    Expression negation = new ArithmeticUnaryExpression(MINUS, column);
                    return p.project(
                            Assignments.of(
                                    // The column reference is part of both the assignments
                                    p.symbol("column_ref", BIGINT), column,
                                    p.symbol("negated_column_ref", BIGINT), negation),
                            p.tableScan(
                                    table,
                                    ImmutableList.of(p.symbol("just_bigint", BIGINT)),
                                    ImmutableMap.of(p.symbol("just_bigint", BIGINT), bigintColumn)));
                })
                .matches(project(
                        ImmutableMap.of(
                                "column_ref", expression("just_bigint_0"),
                                "negated_column_ref", expression("- just_bigint_0")),
                        tableScan(
                                hiveTable.withProjectedColumns(ImmutableSet.of(bigintColumn))::equals,
                                TupleDomain.all(),
                                ImmutableMap.of("just_bigint_0", bigintColumn::equals))));
```

Reply:
thanks, fixed.